core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -746,6 +766,37 @@ private Consumer<Set<String>> failedShareAcknowledgeMetricsHandler() {
         };
     }
 
+    /**
+     * The handler for share version feature metadata changes.
+     * @param shareVersion the new share version feature
+     */
+    public void onShareVersionToggle(ShareVersion shareVersion) {
+        if (!shareVersion.supportsShareGroups()) {
+            isShareGroupsSupported.set(false);
+            // Remove all share sessions from share session cache.
+            synchronized (cache) {
+                cache.removeAllSessions();
+            }
+            Set<SharePartitionKey> sharePartitionKeys = new HashSet<>(partitionCacheMap.keySet());
+            // Remove all share partitions from partition cache.
+            sharePartitionKeys.forEach(sharePartitionKey ->
+                removeSharePartitionFromCache(sharePartitionKey, partitionCacheMap, replicaManager)
+            );
+        } else {
+            isShareGroupsSupported.set(true);
+        }
+    }
+
+    // Visible for testing.
+    protected Map<SharePartitionKey, SharePartition> partitionCacheMap() {
```

Comment:
It seems to me that the tests only really want to know the number of share-partitions in the cache. I wonder if you could have a method that exposes that instead, rather than constructing a hashmap from the internal map.